### DISCUSSION TITLE
Difference between ImmediateRenderer and DeferredRenderer

### DISCUSSION THREAD

**Discussion by DrunkyBard** [AvaloniaUI/Avalonia#5845] [Q&A]
Hi,

I would like to integrate Avalonia application with not .NET app. For that I'm considering to use some functionality from RemoteServerTopLevelImpl class. The thing is that I'm not quite understand the purpose of DeferredRenderer. 
I see that it schedules draw operations in UI thread but actual rendering is performed in non-UI thread. Except that what is the conceptual difference between ImmediateRenderer and DeferredRenderer?

**Comment by kekekeks** [comment]
DeferredRenderer provides better hit-testing for complex shapes since it builds a scene graph. I. e. consider a following control:

```
    public class MyControl : Control
    {
        public override void Render(DrawingContext context)
        {
            context.DrawGeometry(
                Brushes.Aqua, null,
                new EllipseGeometry(new Rect(20,20,20,20)));
            base.Render(context);
        }
    }
```

For ImmediateRenderer that control would be a simple rectangle. Deferred renderer would know that it only has an ellipse partially covering it. So with DeferredRenderer the control would receive mouse events only when mouse is over the ellipse.

**Reply by DrunkyBard** [reply]
Thanks a lot for the clarification.

I would like to use DeferredRenderer for rendering remote application. Could you please advice how to proceed with that?
I've made a rough prototype how it could work ([link](https://github.com/DrunkyBard/Avalonia/commit/dc1b5b4328dc337efd6f4177f58994393b937b8e)). Would you mind to check whether it's the right direction?
In the prototype RemoteServer subscribes to SceneInvalidated event and invalidates the state of RemoteServerTopLevelImpl. 
Once a scene became invalid, a new frame is sent back to RemoteWidget.
FramebufferRenderTarget is used for remote rendering as a frame storage and it's instance is reused for each user input and invalidated each time window requests resize.
Is there any pitfalls that could occur with the given approach?
